### PR TITLE
Use less pageserver-specific method in RemoteStorage trait

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Presumably last PR, stripped off the archiving impl in the forthcoming PR.

Instead of requiring to have a method that returns a very pageserver-specific `RemoteFileInfo`, instead now return a "local path" as the main part of that info. The rest of the info can be derived in-place and dose not have to stuck in the public storage API.

### CODE REVIEW

File: pageserver/src/remote_storage/local_fs.rs [hunk 1]
```diff
@@ -141,9 +134,12 @@ impl RemoteStorage for LocalFs {
...
             );
-            io::copy(&mut source, to)
-                .await
-                .context("Failed to download a file from the local storage")?;
+            io::copy(&mut source, to).await.with_context(|| {
+                format!(
+                    "Failed to download file '{}' from the local storage",
+                    file_path.display()
+                )
+            })?;
```
Comment (patins) [inline pageserver/src/remote_storage/local_fs.rs:142]: Do we also need to flush here

File: pageserver/src/remote_storage/storage_sync.rs [hunk 2]
```diff
@@ -365,19 +368,41 @@ async fn fetch_existing_uploads<P: std::fmt::Debug, S: 'static + RemoteStorage<S
...
+                Ok(path) => path,
+                Err(e) => return Err((e, remote_path)),
+            };
+            let metadata = if local_path
+                .file_name()
+                .and_then(|os_str| os_str.to_str())
+                .unwrap_or_default()
+                == METADATA_FILE_NAME
+            {
+                let mut metadata_bytes = std::io::Cursor::new(Vec::new());
```
Comment (patins) [inline pageserver/src/remote_storage/storage_sync.rs:381]: I think you don't need cursor here

File: pageserver/src/remote_storage/storage_sync.rs [hunk 3]
```diff
@@ -365,19 +368,41 @@ async fn fetch_existing_uploads<P: std::fmt::Debug, S: 'static + RemoteStorage<S
     let mut data_fetches = uploaded_files
         .into_iter()
         .map(|remote_path| async {
-            (
-                remote_file_info(remote_storage, &remote_path).await,
-                remote_path,
-            )
+            let local_path = match remote_storage.local_path(&remote_path) {
```
Comment (patins) [inline pageserver/src/remote_storage/storage_sync.rs:371]: It'd be cool to reduce the duplication of map_err in this function. Maybe use a helper function and `?`.

Comment (SomeoneToIgnore) [reply pageserver/src/remote_storage/storage_sync.rs:371]: Agree, but let's omit that in this PR: the code for fetching remote files is different with archives, so does not make too much sense to make this version readable.

**Review by patins — APPROVED** [review]

**Pull request merged**